geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/WanCopyRegionFunction.java

```diff
@@ -0,0 +1,480 @@
+              senderId));
+    }
+
+    if (!sender.isRunning()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__SENDER__NOT__RUNNING, senderId));
+    }
+
+    if (!sender.isParallel() && !((InternalGatewaySender) sender).isPrimary()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__SENDER__SERIAL__AND__NOT__PRIMARY,
+              senderId));
+    }
+
+    try {
+      return executeWanCopyRegionFunctionInNewThread(context, region, regionName, sender, maxRate,
+          batchSize);
+    } catch (InterruptedException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.WAN_COPY_REGION__MSG__EXECUTION__CANCELED);
+    } catch (ExecutionException e) {
+      Writer buffer = new StringWriter();
+      PrintWriter pw = new PrintWriter(buffer);
+      e.printStackTrace(pw);
+      logger.error("Exception when running wan-copy region command: {}", buffer.toString());
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__EXECUTION__FAILED, e.getMessage()));
+    }
+  }
+
+  private CliFunctionResult executeWanCopyRegionFunctionInNewThread(
+      FunctionContext<Object[]> context,
+      Region region, String regionName, GatewaySender sender, long maxRate, int batchSize)
+      throws InterruptedException, ExecutionException {
+    ExecutorService executor = LoggingExecutors
+        .newSingleThreadExecutor(getFunctionThreadName(regionName,
+            sender.getId()), true);
+    Callable<CliFunctionResult> callable =
+        new wanCopyRegionCallable(context, region, sender, maxRate, batchSize);
+    Future<CliFunctionResult> future = executor.submit(callable);
+    return future.get();
+  }
+
+  class wanCopyRegionCallable implements Callable<CliFunctionResult> {
+    private final FunctionContext<Object[]> context;
+    private final Region region;
+    private final GatewaySender sender;
+    private final long maxRate;
+    private final int batchSize;
+
+    public wanCopyRegionCallable(final FunctionContext<Object[]> context, final Region region,
+        final GatewaySender sender, final long maxRate,
+        final int batchSize) {
+      this.context = context;
+      this.region = region;
+      this.sender = sender;
+      this.maxRate = maxRate;
+      this.batchSize = batchSize;
+    }
+
+    @Override
+    public CliFunctionResult call() throws Exception {
+      return wanCopyRegion(context, region, sender, maxRate, batchSize);
+    }
+  }
+
+  @VisibleForTesting
+  CliFunctionResult wanCopyRegion(FunctionContext<Object[]> context, Region region,
```

Comment:
Can you please extract some smaller methods from this? There's actually a maximum byte length for the size of methods in order for the Jit to perform inlining and other optimizations. Basically, if you carve it up into smaller methods, then Jit can do inlining and it will actually perform better.

One way to do this is to extract blocks as methods to avoid having nested try-blocks and while-blocks. So if you see a nested while-block, pull that out as its own method. You can lean on IntelliJ refactoring to make it easy.

Reply:
I have done it at the cost of creating a new static class holding the connection status.
But I think it is cleaner now.